src/buffer/out/textBuffer.cpp

```diff
@@ -2888,125 +2900,400 @@ std::vector<til::point_span> TextBuffer::SearchText(const std::wstring_view& nee
-    const auto& start{ curr.end };
-    const auto& end{ GetCursor().GetPosition() };
+void TextBuffer::StartPrompt()
+{
+    const auto currentRowOffset = GetCursor().GetPosition().y;
+    auto& currentRow = GetMutableRowByOffset(currentRowOffset);
+    currentRow.StartPrompt();
 
-    const auto line = start.y;
-    const auto& row = GetRowByOffset(line);
-    return row.GetText(start.x, end.x);
+    auto attr = GetCurrentAttributes();
+    attr.SetMarkAttributes(MarkKind::Prompt);
+    SetCurrentAttributes(attr);
 }
 
-void TextBuffer::SetCurrentPromptEnd(const til::point pos) noexcept
+bool TextBuffer::_createPromptMarkIfNeeded()
 {
-    if (_marks.empty())
+    // We might get here out-of-order, without seeing a StartPrompt (FTCS A)
+    // first. Since StartPrompt actually sets up the prompt mark on the ROW, we
+    // need to do a bit of extra work here to start a new mark (if the last one
+    // wasn't in an appropriate state).
+
+    const auto mostRecentMarks = GetMarkExtents(1u);
+    if (!mostRecentMarks.empty())
     {
-        return;
+        const auto& mostRecentMark = til::at(mostRecentMarks, 0);
+        if (mostRecentMark.HasOutput())
+        {
+            // The most recent command mark had output. That suggests that either:
+            // * shell integration wasn't enabled (but the user would still
+            //   like lines with enters to be marked as prompts)
+            // * or we're in the middle of a command that's ongoing.
+
+            // If the mark doesn't have any command - then we know we're
+            // playing silly games with just marking whole lines as prompts,
+            // then immediately going to output.
+            //   --> add a new mark to this row, set all the attrs in this
+            //   row to be Prompt, and set the current attrs to Output.
+            //
+            // If it does have a command, then we're still in the output of
+            // that command.
+            //   --> the current attrs should already be set to Output.
+            if (!mostRecentMark.HasCommand())
+            {
+                auto& row = GetMutableRowByOffset(GetCursor().GetPosition().y);
+                row.StartPrompt();
+                return true;
+            }
+        }
+        else
+        {
+            // The most recent command mark _didn't_ have output yet. Great!
+            // we'll leave it alone, and just start treating text as Command or Output.
+        }
     }
-    auto& curr{ _marks.back() };
-    curr.end = pos;
-}
-void TextBuffer::SetCurrentCommandEnd(const til::point pos) noexcept
-{
-    if (_marks.empty())
+    else
     {
-        return;
+        // There were no marks at all!
+        //   --> add a new mark to this row, set all the attrs in this row
+        //   to be Prompt, and set the current attrs to Output.
+
+        auto& row = GetMutableRowByOffset(GetCursor().GetPosition().y);
+        row.StartPrompt();
+        return true;

```

Comment:
this is shared with R3253 above - can we flip the conditions around so that we can early-return if we're not gonna do anything, and then consolidate this logic at the bottom? or is it not that simple?

Reply:
Done